parser.rst

```diff
@@ -0,0 +1,901 @@
+cost comes for continuously checking if the given rule has a cache hit or not. In many situations, just parsing it
+again can be faster. Pegen **disables memoization by default** except for rules with the special marker `memo` after
+the rule name (and type, if present): ::
+
+    rule_name[typr] (memo):
+        ...
+
+Selectively turning on memoization for a handful of rules, the parser becomes faster and uses less memory.
+
+.. note::
+    Left-recursive rules always use memoization, since the implementation of left-recursion depends on it.
+
+To know if a new rule needs memoization or not, benchmarking is required
+(comparing execution times and memory usage of some considerably big files with
+and without memoization). There is a very simple instrumentation API available
+in the C generated parse code that allows to measure how much each rule uses
+memoization (check the :file:`Parser/pegen.c` file for more information) but it
+needs to be manually activated.
+
+Automatic variables
+~~~~~~~~~~~~~~~~~~~
+
+To make writing actions easier, Pegen inject some automatic variables in the namespace available
+when writing actions. In the C parser, some of these automatic variable names are:
+
+* ``p``: The parser structure.
+* ``EXTRA``: This is a macro that expands to ``(_start_lineno, _start_col_offset, _end_lineno, _end_col_offset, p->arena)``,
+  which is normally used to create AST nodes as almost all constructors need these attributes to be provided. All of the
+  location variables are taken from the location information of the current token.
+
+Hard and Soft keywords
+~~~~~~~~~~~~~~~~~~~~~~
+
+.. note::
+    In the grammar files, keywords are defined using **single quotes** (e.g. `'class'`) while soft
+    keywords are defined using **double quotes** (e.g. `"match"`).
+
+There are two kinds of keywords allowed in pegen grammars: *hard* and *soft*
+keywords. The difference between hard and soft keywords is that hard keywords
+are always reserved words, even in positions where they make no sense (e.g. ``x = class + 1``),
+while soft keywords only get a special meaning in context. Trying to use a hard
+keyword as a variable will always fail:
+
+.. code-block::
+
+    >>> class = 3
+    File "<stdin>", line 1
+        class = 3
+            ^
+    SyntaxError: invalid syntax
+    >>> foo(class=3)
+    File "<stdin>", line 1
+        foo(class=3)
+            ^^^^^
+    SyntaxError: invalid syntax
+
+While soft keywords don't have this limitation if used in a context other when they
+are defined as keywords:
+
+.. code-block:: python
+
+    >>> match = 45
+    >>> foo(match="Yeah!")
+
+The ``match`` and ``case`` keywords are soft keywords, so that they are recognized as
+keywords at the beginning of a match statement or case block respectively, but are
+allowed to be used in other places as variable or argument names.
+
+You can get a list of all keywords defined in the grammar from Python:
+
+.. code-block:: python
+
+    >>> keyword.kwlist
```

Comment:
It might be obvious, but what about adding an import in this code-block?

Reply:
👍